### ISSUE TITLE
stack overflow when trying to compile this code snippet

### ISSUE THREAD

**Issue by flo-l** [rust-lang/rust#22773]
I'm on Fedora 21, x86_64.

```rust
pub struct A<I: Iterator> {
    iter: I,
    buf: Vec<I::Item>,
}

impl<I: Iterator> Iterator for A<I> where
    I: Iterator<Item=I::Item>
{
    fn next(&mut self) -> Option<I::Item> {
        None
    }
}

fn main() {}
```

Output:

```
$ rustc test.rs 

thread 'rustc' has overflowed its stack
Ungültiger Maschinenbefehl (Speicherabzug geschrieben)
```

(Sorry for the german :laughing: )

You can also watch it fail with the [playpen](http://is.gd/LjFzf1)!

**Comment by steveklabnik** [comment]
not strictly an ICE but close enough

**Comment by flo-l** [comment]
probably fixed by #22512, the commit message says it fixes a few stack overflows related to associated types. I'll recheck if this still crashes in two days or so when the patch is the nightly! ;)

**Comment by tamird** [comment]
No longer ICEs.
```shell
$ rustc main.rs
main.rs:7:22: 7:29 error: unsupported cyclic reference between types/traits detected
main.rs:7     I: Iterator<Item=I::Item>
                               ^~~~~~~
note: the cycle begins when computing the bounds for type parameter `I`...
note: ...which then again requires computing the bounds for type parameter `I`, completing the cycle.
error: aborting due to previous error
```
